bridges/modules/xcm-bridge/src/congestion.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The module contains utilities for handling congestion between the bridge hub and routers.
+
+use crate::{Bridges, Config, DispatchChannelStatusProvider, LOG_TARGET};
+use bp_xcm_bridge::{BridgeId, LocalXcmChannelManager, Receiver};
+use codec::{Decode, Encode};
+use sp_runtime::traits::Convert;
+use sp_std::{marker::PhantomData, vec::Vec};
+use xcm::latest::{send_xcm, Location, SendXcm, Xcm};
+use xcm_builder::{DispatchBlob, DispatchBlobError};
+
+/// Limits for handling congestion.
+#[derive(Debug, Decode, Encode)]
+pub struct CongestionLimits {
+	/// Maximal number of messages in the outbound bridge queue. Once we reach this limit, we
+	/// suspend a bridge.
+	pub outbound_lane_congested_threshold: bp_messages::MessageNonce,
+	/// After we have suspended the bridge, we wait until number of messages in the outbound bridge
+	/// queue drops to this count, before sending resuming the bridge.
+	pub outbound_lane_uncongested_threshold: bp_messages::MessageNonce,
+	/// Maximal number of messages in the outbound bridge queue after we have suspended the bridge.
+	/// Once we reach this limit, we stop exporting more messages.
+	pub outbound_lane_stop_threshold: bp_messages::MessageNonce,
+}
+
+impl CongestionLimits {
+	/// Checks if limits are valid.
+	pub fn is_valid(&self) -> bool {
+		self.outbound_lane_uncongested_threshold < self.outbound_lane_congested_threshold &&
+			self.outbound_lane_stop_threshold > self.outbound_lane_congested_threshold
```

Comment:
Lokking at the XCMP `QueueConfigData::validate()`, I think we should also check that `outbound_lane_uncongested_threshold > 0`. Otherwise we will never resume.

And on a more high level note, can we reuse the XCMP `QueueConfigData` here ? Looks like it contains the same fields, only under different names.

Reply:
> Looking at the XCMP `QueueConfigData::validate()`, I think we should also check that `outbound_lane_uncongested_threshold > 0`. Otherwise we will never resume.

thank you, I will check the condition and also if I can easily split this PR to a smaller pieces.

> And on a more general note, can we reuse the XCMP `QueueConfigData` here ? Looks like it contains the same fields, only under different names.

`QueueConfigData` is defined in `cumulus-pallet-xcmp-queue`, so I would add it as a dependency or where would you suggest to move `QueueConfigData`?

Reply:
> `QueueConfigData` is defined in `cumulus-pallet-xcmp-queue`, so I would add it as a dependency or where would you suggest to move `QueueConfigData`?

I think it would be best to move it to a primitives crate . Not sure which one would be the best of if a new one would be needed.

Reply:
> > `QueueConfigData` is defined in `cumulus-pallet-xcmp-queue`, so I would add it as a dependency or where would you suggest to move `QueueConfigData`?
> 
> I think it would be best to move it to a primitives crate . Not sure which one would be the best of if a new one would be needed.

I didn't want to do too much major/breaking stuff here, but I think this is a good idea to reuse `QueueConfigData` as you suggest.

If we move `QueueConfigData` from the xcmp-queue, it will be probably `major` for SemVer, maybe some re-export/alias in xcmp-queue would work for minor?  `pub struct QueueConfigData` -> `pub type QueueConfigData = some_primitives_module::QueueConfigData`?

@serban300 you can do it on the top of this PR or separate: https://github.com/paritytech/polkadot-sdk/issues/8463